### PR TITLE
Add DQM solver

### PR DESCRIPTION
**Pull request by randomir**
- abstract `dwave.cloud.solver.UnstructuredSolver` to `.BaseUnstructuredSolver`
- alias `dwave.cloud.solver.BQMSolver` as `.UnstructuredSolver`
- add `dwave.cloud.solver.DQMSolver`, with `sample_dqm` method, as well as fallback `sample_{bqm,ising,qubo}`

### CODE REVIEW

File: dwave/cloud/solver.py [hunk 1]
```diff
@@ -408,16 +499,93 @@ def upload_bqm(self, bqm):
...
-            # assume `bqm` given as file, ready for upload
-            data = bqm
+            # assume `dqm` given as file, ready for upload
+            data = dqm
 
-        return self.client.upload_problem_encoded(data)
+        logger.debug("Problem encoded for upload: %r", data)
+        return data
+
+    def sample_bqm(self, bqm, **params):
```
Comment (arcondello) [inline dwave/cloud/solver.py:541]: Does this need to be "public"? Just not necessarily something that we want to encourage folks to do

**Review by arcondello — APPROVED** [review]

File: dwave/cloud/solver.py (hunk 1 above)
Comment (randomir) [reply dwave/cloud/solver.py:541]: Not sure what you mean by "this". `sample_bqm` method?

Comment (randomir) [reply dwave/cloud/solver.py:541]: Oh, you mean on the `DQMSolver`. Well, to make the new type better fit in, for now, yes.

Comment (arcondello) [reply dwave/cloud/solver.py:541]: I wonder if maybe a doc string that says "it's better to use a BQM solver" or similar would be good.

Comment (randomir) [reply dwave/cloud/solver.py:541]: Btw, I would not document this method, that's for sure. Like I wouldn't *many* other "public" methods in this module. /cc @JoelPasvolsky

File: dwave/cloud/solver.py [hunk 2]
```diff
@@ -345,49 +377,108 @@ def _encode_problem_as_ref(self, problem, params):
...
+            solver=self, id_=None, return_matrix=self.return_matrix)
 
         logger.debug("Submitting new problem to: %s", self.id)
         self.client._submit(body, computation)
 
         return computation
 
+
+class BQMSolver(BaseUnstructuredSolver):
+    """Class for D-Wave unstructured binary quadratic model solvers.
```
Comment (JoelPasvolsky) [inline dwave/cloud/solver.py:432]: "Class for D-Wave unstructured" --> I think we generally refer to these as Leap rather than D-Wave

**Review by JoelPasvolsky — COMMENTED** [review]
"handled" is more standardly "supported"

**Pull request merged**